sdks/python/apache_beam/utils/windowed_value.py

```diff
@@ -279,6 +293,208 @@ def create(value, timestamp_micros, windows, pane_info=PANE_INFO_UNKNOWN):
+  """
+  def __init__(
+      self,
+      values,
+      timestamps,  # type: Sequence[TimestampTypes]
+      windows,  # type: Iterable[Tuple[BoundedWindow, ...]]
+      pane_infos=PANE_INFO_UNKNOWN  # type: Union[Iterable[PaneInfo],PaneInfo]
+  ):
+    self.values = values
+
+    def convert_timestamp(timestamp: TimestampTypes) -> int:
+      if isinstance(timestamp, int):
+        return timestamp * 1000000
+      else:
+        # TODO: Cache Timestamp object as in WindowedValue?
+        timestamp_object = (
+            timestamp
+            if isinstance(timestamp, Timestamp) else Timestamp.of(timestamp))
+        return timestamp_object.micros
+
+    self.timestamp_objects: Optional[List[Timestamp]] = None
+    self.timestamps_micros = [convert_timestamp(t) for t in timestamps]
+    self.windows = windows
+    #TODO: Should we store length?
+    #self.length = length
+    self.pane_infos = pane_infos
+
+  @property
+  def timestamps(self) -> Sequence[Timestamp]:
+    if self.timestamp_objects is None:
+      self.timestamp_objects = [
+          Timestamp(0, micros) for micros in self.timestamps_micros
+      ]
+
+    return self.timestamp_objects
+
+  def with_values(self, new_values):
+    # type: (Any) -> WindowedBatch
+
+    """Creates a new WindowedBatch with the same timestamps and windows as this.
+
+    This is the fasted way to create a new WindowedValue.
+    """
+    return create_batch(
+        new_values, self.timestamps_micros, self.windows, self.pane_infos)
+
+  def as_windowed_values(self, explode_fn: Callable) -> Iterable[WindowedValue]:
+    for value, timestamp, windows, pane_info in zip(explode_fn(self.values),
+                                                    self.timestamps_micros,
+                                                    self.windows,
+                                                    self._pane_infos_iter()):
+      yield create(value, timestamp, windows, pane_info)
+
+  def _pane_infos_iter(self):
+    if isinstance(self.pane_infos, PaneInfo):
+      return itertools.repeat(self.pane_infos, len(self.timestamps_micros))
+    else:
+      return self.pane_infos
+
+  def __eq__(self, other):
+    if isinstance(other, ConcreteWindowedBatch):
+      return (
+          type(self) == type(other) and
+          self.timestamps_micros == other.timestamps_micros and
+          self.values == other.values and self.windows == other.windows and
+          self.pane_infos == other.pane_infos)
+    return NotImplemented
+
+  def __hash__(self):
+    if isinstance(self.pane_infos, PaneInfo):
+      pane_infos_hash = hash(self.pane_infos)
+    else:
+      pane_infos_hash = sum(hash(p) for p in self.pane_infos)
+
+    return ((hash(self.values) & 0xFFFFFFFFFFFFFFF) + 3 *
+            (sum(self.timestamps_micros) & 0xFFFFFFFFFFFFFF) + 7 *
+            (sum(hash(w) for w in self.windows) & 0xFFFFFFFFFFFFF) + 11 *
+            (pane_infos_hash & 0xFFFFFFFFFFFFF))
+
+
+def create_batch(
```

Comment:
This indirection was for performance reasons; just inline it above.

Reply:
Done